### PR TITLE
197 198 199 multi project sessions

### LINKED ISSUE #197: Jira Authenticator: Holt Liste mit Projekten, für die gegebene Credentials valide sind
**User Story**
Technisches Requirement

**Acceptance Criteria**
- Wenn der Jira Authenticator Credentials bekommt, frägt er beim Jira-Server an, für welche Projekte der Nutzer Schreibrechte hat.

### LINKED ISSUE #198: Jira Authenticator: Prüft, ob das Projekt, für das sich der Benutzer anmelden will, auch in der Liste ist.
**User Story**
Technisches Requirement

**Acceptance Criteria**
- Der Authenticator prüft, ob das Projekt, für das sich der Nutzer anmelden will, in der Liste der Projekte ist, für das der Nutzer Leserechte hat.

### LINKED ISSUE #199: Erweiterung der "User-Session und des Projekt-Cookies", um die Liste aller für den Nutzer zugänglichen Projekte
**User Story**
Technisches Requirement

**Acceptance Criteria**
- Die `UserSession` (sh. Ecore) enthält nicht nur die Information für ein Projekt, sondern für mehrere. Insb. das allowedPathPattern muss erweitert werden, sodass Zugriff auf mehrere Projekte möglich ist.
- Der `User` (sh. Ecore) enthält eine Liste von Projekten, für die der Nutzer Zugriff hat.
- Der Cookie, welcher im `AuthRestServiceBase` die Projekte speichert, für die ein Nutzer Zugriff hat, enthält eine Liste der Projekte, auf die der Nutzer Zugriff hat.

### PR DESCRIPTION
**Pull request by hauptmab**
Closes #197
Closes #198
Closes #199

### CODE REVIEW

File: bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java [hunk 1]
```diff
@@ -50,7 +52,7 @@ private static void configureInMemorySessionService() throws Exception {
 	public void testIsAuthorized() throws SpecmateException {
```
Comment (junkerm) [inline bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java]: please also add tests for the multi-project case

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java [hunk 2]
```diff
@@ -23,6 +25,20 @@
 	 */
 	IContainer getContainerForRequirement(Requirement requirement) throws SpecmateException;
 
-	boolean authenticate(String username, String password) throws SpecmateException;
+	/**
+	 * Returns a set of projects given credentials have access to. Returns null if
```
Comment (junkerm) [inline bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java]: wouldn't it be better to throw an exception or return an empty list if the credentials are invalid instead of returning null?

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java [hunk 3]
```diff
@@ -23,6 +25,20 @@
...
+	 * @param project        The project the user wants to login
+	 * @param projectService The project service (helpful to retrieve other
+	 *                       projects).
+	 * @return Returns a set of projects the credentials can access. If succesfull,
+	 *         this set must contain the project the user tries to login.
+	 *         Furthermore, this set may contain other projects, the user can also
+	 *         access. E.g., a user log on to a jira-project and is logged in also
+	 *         to other jira-projects his account can also access.
+	 */
+	Set<IProject> authenticate(String username, String password, IProject project, IProjectService projectService)
```
Comment (junkerm) [inline bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java]: It should not be necessary to pass the project service as parameter, as the implementing classes will be components which can get the project service injected

File: bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java [hunk 4]
```diff
@@ -22,27 +27,48 @@
...
+	
 	@Override
 	public UserSession authenticate(String username, String password, String projectname) throws SpecmateException {
 		IProject project = projectService.getProject(projectname);
 		if (project == null) {
 			throw new SpecmateAuthorizationException("Project " + projectname + " does not exist.");
 		}
-		boolean authenticated = project.getConnector().authenticate(username, password);
-		if (!authenticated) {
+		Set<IProject> authenticatedProjects = project.getConnector().authenticate(username, password, project, projectService);
```
Comment (junkerm) [inline bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java]: passing the project service should not be necessary as the connectors can get the project service injected

File: bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java [hunk 5]
```diff
@@ -22,27 +27,48 @@
...
 		} catch (SpecmateException e) {
 			// Something went wrong when creating the session. We act as if the
 			// Authorization failed
 			throw new SpecmateAuthorizationException("Could not create a user session. Reason: " + e.getMessage(), e);
 		}
 	}
+	
+	private Set<String> getProjectNames(Set<IProject> projects) {
+
+		Set<String> projectNames = new HashSet<>();
```
Comment (junkerm) [inline bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java]: with IProject.getID() you can access the project name. So probably you can shorten this code to
projects.stream().map(p->p.getName()).collect(Collectors.toSet)

File: bundles/specmate-config/config/specmate-config.properties [hunk 6]
```diff
@@ -51,7 +51,7 @@ search.maxResults 		= 100
 
 # Projects
 ## List of project ids, each project listed here needs to be configured below
-#project.projects = 
+project.projects=jiraMFP,jiraMSP
```
Comment (junkerm) [inline bundles/specmate-config/config/specmate-config.properties]: please remove configuration changes

File: bundles/specmate-config/config/specmate-config.properties [hunk 7]
```diff
@@ -91,3 +91,24 @@ session.persistent			= true
 ## NLP Service
 nlp.spacy.url = http://127.0.0.1:80
 nlp.spacy.model = en_core_web_lg
+
```
Comment (junkerm) [inline bundles/specmate-config/config/specmate-config.properties]: please remove ocnfiguration chagnes

File: bundles/specmate-file-connector/src/com/specmate/connectors/fileconnector/internal/FileConnector.java [hunk 8]
```diff
@@ -171,11 +176,11 @@ public IContainer getContainerForRequirement(Requirement requirement) {
...
 	@Override
-	public boolean authenticate(String username, String password) {
-		if (user == null) {
-			return false;
+	public Set<IProject> authenticate(String username, String password, IProject logonProject, IProjectService projectService) {
+		if (username.equals(user) && password.equals(this.password)) {
+			return new HashSet<IProject>(Arrays.asList(logonProject));
 		} else {
-			return username.equals(user) && password.equals(this.password);
+			return null;
```
Comment (junkerm) [inline bundles/specmate-file-connector/src/com/specmate/connectors/fileconnector/internal/FileConnector.java]: See comment above, I would prefer explicit handling by an exception

File: bundles/specmate-hp-connector/src/com/specmate/connectors/hpconnector/internal/services/HPConnector.java [hunk 9]
```diff
@@ -172,8 +177,13 @@ public void setLogService(LogService logService) {
...
 
 	@Override
-	public boolean authenticate(String username, String password) throws SpecmateException {
-		return this.hpConnection.authenticateRead(username, password, this.hpProjectName);
+	public Set<IProject> authenticate(String username, String password, IProject logonProject, IProjectService projectService) throws SpecmateException {		
+		
+		if (this.hpConnection.authenticateRead(username, password, this.hpProjectName)) {
+			return new HashSet<IProject>(Arrays.asList(logonProject));
+		} else {
+			return null;
```
Comment (junkerm) [inline bundles/specmate-hp-connector/src/com/specmate/connectors/hpconnector/internal/services/HPConnector.java]: see comments above

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java [hunk 10]
```diff
@@ -341,8 +345,36 @@ public IContainer getContainerForRequirement(Requirement requirement) throws Spe
 	}
 
 	@Override
-	public boolean authenticate(String username, String password) throws SpecmateException {
-		return JiraUtil.authenticate(url, projectName, username, password);
+	public Set<IProject> authenticate(String username, String password, IProject logonProject, IProjectService projectService) throws SpecmateException {
```
Comment (junkerm) [inline bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java]: get the project servvice injected using @Reference

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java [hunk 11]
```diff
@@ -341,8 +345,36 @@ public IContainer getContainerForRequirement(Requirement requirement) throws Spe
...
 
 	@Override
-	public boolean authenticate(String username, String password) throws SpecmateException {
-		return JiraUtil.authenticate(url, projectName, username, password);
+	public Set<IProject> authenticate(String username, String password, IProject logonProject, IProjectService projectService) throws SpecmateException {		
+		
+		List<String> accessibleJiraProjectNames = JiraUtil.getProjects(url, username, password);
+		Set<IProject> accessibleSpecmateProjectNames = new HashSet<>();
+		
+		if (! (logonProject.getConnector() instanceof JiraConnector)) {
```
Comment (junkerm) [inline bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java]: as discussed: remove first part of the method and eliminate logonProject and projectService parameter

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java [hunk 12]
```diff
@@ -28,6 +37,7 @@ public static boolean authenticate(String url, String project, String username,
 		try {
 			client = JiraUtil.createJiraRESTClient(url, username, password);
 			client.getProjectClient().getProject(project).claim();
+
```
Comment (junkerm) [inline bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java]: remove empty line

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java [hunk 13]
```diff
@@ -49,6 +59,42 @@ public static boolean authenticate(String url, String project, String username,
...
+			if (status == 401) {
+				return null;
+			}
+			return null;
+		} finally {
+			if (client != null) {
+				try {
+					client.close();
+				} catch (IOException e) {
+					// an error occured - better return false
```
Comment (junkerm) [inline bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java:90]: update comment

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java [hunk 14]
```diff
@@ -49,6 +59,42 @@ public static boolean authenticate(String url, String project, String username,
...
+			}
+
+			return projectKeys;
+
+		} catch (URISyntaxException e) {
+			throw new SpecmateAuthorizationException("Jira authentication failed.", e);
+		} catch (RestClientException e) {
+			Integer status = e.getStatusCode().get();
+			if (status == 401) {
+				return null;
```
Comment (junkerm) [inline bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java:82]: see other comments - i would avoid returning null

File: bundles/specmate-std-env/dev-specmate-all.bndrun [hunk 15]
```diff
@@ -226,7 +226,7 @@
 
 -runproperties: \
 	jetty.http.port=8080,\
-	#jetty.ssl.port=8443,\
+	#jetty.ssl.port=8080,\
```
Comment (junkerm) [inline bundles/specmate-std-env/dev-specmate-all.bndrun]: revert change in run config

**Review by junkerm — CHANGES_REQUESTED** [review]
see my comments

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java (hunk 2 above)
Comment (hauptmab) [reply bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java]: Good point. However, I do have issues throwing exceptions for expectable cases (such as wrong credentials, which are no exceptions). An empty set is by far better than null.

I updated the documentation and removed null-check

File: bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java (hunk 3 above)
Comment (hauptmab) [reply bundles/specmate-connectors-api/src/com/specmate/connectors/api/IRequirementsSource.java]: You are absolutely right. the project object is now accessible directly for each connector. The parameter is removed from the method signature.

File: bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java (hunk 4 above)
Comment (hauptmab) [reply bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java]: I like the idea. However, I have issues finding the right spot for injecting the project service. Do you have a recommentation @junkerm ?

File: bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java (hunk 5 above)
Comment (hauptmab) [reply bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java]: good point. I made the recommended change.

File: bundles/specmate-config/config/specmate-config.properties (hunk 6 above)
Comment (hauptmab) [reply bundles/specmate-config/config/specmate-config.properties]: sorry :-o

File: bundles/specmate-config/config/specmate-config.properties (hunk 7 above)
Comment (hauptmab) [reply bundles/specmate-config/config/specmate-config.properties]: sorry :-o

File: bundles/specmate-file-connector/src/com/specmate/connectors/fileconnector/internal/FileConnector.java (hunk 8 above)
Comment (hauptmab) [reply bundles/specmate-file-connector/src/com/specmate/connectors/fileconnector/internal/FileConnector.java]: done (see above)

File: bundles/specmate-hp-connector/src/com/specmate/connectors/hpconnector/internal/services/HPConnector.java (hunk 9 above)
Comment (hauptmab) [reply bundles/specmate-hp-connector/src/com/specmate/connectors/hpconnector/internal/services/HPConnector.java]: done (see above)

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java (hunk 12 above)
Comment (hauptmab) [reply bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java]: done

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java (hunk 13 above)
Comment (hauptmab) [reply bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java:90]: (see other pull request)

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java (hunk 14 above)
Comment (hauptmab) [reply bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraUtil.java:82]: (see other pull request)

File: bundles/specmate-std-env/dev-specmate-all.bndrun (hunk 15 above)
Comment (hauptmab) [reply bundles/specmate-std-env/dev-specmate-all.bndrun]: done

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java (hunk 10 above)
Comment (hauptmab) [reply bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java]: done

File: bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java (hunk 11 above)
Comment (hauptmab) [reply bundles/specmate-jira-connector/src/com/specmate/connectors/jira/internal/services/JiraConnector.java]: dome

File: bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java (hunk 1 above)
Comment (hauptmab) [reply bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java]: add and improved tests

File: bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java (hunk 4 above)
Comment (hauptmab) [reply bundles/specmate-auth/src/com/specmate/auth/internal/AuthenticationServiceImpl.java]: I found a good way.

File: bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java [hunk 16]
```diff
@@ -49,14 +49,67 @@ private static void configureInMemorySessionService() throws Exception {
...
+	public void testIsNotAuthorized() throws SpecmateException {
+		String projectName1 = "project1";
+
 		UserSession session = sessionService.create(AccessRights.ALL, AccessRights.ALL, userName, password,
-				new HashSet<String>(Arrays.asList(projectName)));
-		assertTrue(sessionService.isAuthorizedPath(session.getId(), baseURL + projectName + "/resource1"));
-		assertTrue(sessionService.isAuthorizedPath(session.getId(), baseURL + projectName + "/resource1/resource2"));
-		assertTrue(sessionService.isAuthorizedPath(session.getId(), baseURL + projectName + "/"));
-		assertFalse(sessionService.isAuthorizedPath(session.getId(), baseURL + projectName));
+				new HashSet<String>());
```
Comment (junkerm) [inline bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java:56]: i would add one additional test to check if authorization is prevented when the list of projects is not empty but contains a wrong project

**Review by junkerm — DISMISSED** [review]
small comment regarding test - otherwise merge

File: bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java (hunk 16 above)
Comment (hauptmab) [reply bundles/specmate-auth-test/src/com/specmate/auth/test/InMemorySessionServiceTest.java:56]: Good point. I added another test case.

**Review by junkerm — APPROVED** [review]
